### PR TITLE
refactor(analysis): rust based page-static-info, deprecate js parse interface in next-swc

### PR DESCRIPTION
**Pull request by kwonoj**
### Description

This PR refactors existing `analysis/get-static-page-info`, moves over most of parse / ast visiting logic into next-swc's rust codebase. By having this, turbopack can reuse same logic to extract info for the analysis. Also as a side effect, this removes JS side parse which is known to be inefficient due to serialization / deserialization.

The entrypoint `getPageStaticInfo` is still in the existing `get-page-static-info`, only for extracting / visiting logic is moved. There are some JS specific context to postprocess extracted information which would require additional effort to move into.

Closes PACK-2088

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| buildDuration | 10.7s | 10.6s | N/A |
| buildDurationCached | 6s | 6.1s | ⚠️ +102ms |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +1.99 kB |
| nextStartRea..uration (ms) | 420ms | 422ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| 170-HASH.js gzip | 30.6 kB | 30.6 kB | N/A |
| 199.HASH.js gzip | 181 B | 185 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 240 B | 241 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 98.5 kB | 98.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| _app-HASH.js gzip | 195 B | 194 B | N/A |
| _error-HASH.js gzip | 183 B | 182 B | N/A |
| amp-HASH.js gzip | 501 B | 501 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 255 B | ✓ |
| head-HASH.js gzip | 349 B | 350 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.6 kB | 2.6 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 309 B | N/A |
| script-HASH.js gzip | 384 B | 384 B | ✓ |
| withRouter-HASH.js gzip | 307 B | 306 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 483 B | ✓ |
| Overall change | 483 B | 483 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 543 B | 541 B | N/A |
| withRouter.html gzip | 523 B | 524 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.5 kB | 93.5 kB | N/A |
| page.js gzip | 146 kB | 146 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 625 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.5 kB | 37.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor-parse-module | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 168 kB | 168 kB | ✓ |
| app-page-exp..prod.js gzip | 93.8 kB | 93.8 kB | ✓ |
| app-page-tur..prod.js gzip | 94.5 kB | 94.5 kB | ✓ |
| app-page-tur..prod.js gzip | 89.1 kB | 89.1 kB | ✓ |
| app-page.run...dev.js gzip | 138 kB | 138 kB | ✓ |
| app-page.run..prod.js gzip | 88.4 kB | 88.4 kB | ✓ |
| app-route-ex...dev.js gzip | 24.2 kB | 24.2 kB | ✓ |
| app-route-ex..prod.js gzip | 16.8 kB | 16.8 kB | ✓ |
| app-route-tu..prod.js gzip | 16.9 kB | 16.9 kB | ✓ |
| app-route-tu..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route.ru...dev.js gzip | 23.6 kB | 23.6 kB | ✓ |
| app-route.ru..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.6 kB | 22.6 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.4 kB | 49.4 kB | ✓ |
| Overall change | 930 kB | 930 kB | ✓ |

</details>

</details>
Commit: fe25ca393deb1870d25581aa889b34d354923a0e

**Comment by ijjk** [comment]
## Failing test suites 
Commit: f7aff08f8130631c609c3c20b6eb6f56a311db21

`pnpm test test/integration/telemetry/test/index.test.js` 
- Telemetry CLI > production mode > emits event when swc fails to load

<details>
<summary>Expand output</summary>

  ● Telemetry CLI › production mode › emits event when swc fails to load

    expect(received).toMatch(expected)

    Expected pattern: /NEXT_SWC_LOAD_FAILURE/
    Received string:  " ⚠ Attempted to load @next/swc-linux-x64-gnu, but it was not installed
     ⚠ Attempted to load @next/swc-linux-x64-musl, but it was not installed
     ⨯ Failed to load SWC binary for linux/x64, see more info here: https://nextjs.org/docs/messages/failed-loading-swc
    "

      138 |         },
      139 |       })
    > 140 |       expect(stderr).toMatch(/NEXT_SWC_LOAD_FAILURE/)
          |                      ^
      141 |       expect(stderr).toContain(
      142 |         `"nextVersion": "${require('next/package.json').version}"`
      143 |       )

      at Object.toMatch (integration/telemetry/test/index.test.js:140:22)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/build/swc/index.ts [hunk 1]
```diff
@@ -1150,14 +1185,33 @@ async function loadWasm(importPath = '') {
...
-          return bindings?.parse
-            ? bindings.parse(src.toString(), options)
-            : Promise.resolve(bindings.parseSync(src.toString(), options))
-        },
-        parseSync(src: string, options: any) {
-          const astStr = bindings.parseSync(src.toString(), options)
-          return astStr
+        analysis: {
+          isDynamicMetadataRoute: async (pageFilePath: string) => {
+            const fileContent = (await tryToReadFile(pageFilePath, true)) || ''
```
Comment (timneutkens) [inline packages/next/src/build/swc/index.ts:1190]: For swc-loader transforms we have the SWC-side read the file, seems we could do the same here in a follow-up PR.

File: packages/next/src/build/analysis/get-page-static-info.ts [hunk 2]
```diff
@@ -483,13 +255,13 @@ export async function getPageStaticInfo(params: {
...
-      fileContent
-    )
-  ) {
-    const swcAST = await parseModule(pageFilePath, fileContent)
+  const binding = await require('../swc').loadBindings()
+  const pageStaticInfo = await binding.analysis.getPageStaticInfo(params)
+
+  if (pageStaticInfo) {
+    const { exportsInfo, extractedValues, rscInfo, warnings } =
+      await binding.analysis.getPageStaticInfo(params)
```
Comment (timneutkens) [inline packages/next/src/build/analysis/get-page-static-info.ts:263]: Seems you're running `binding.analysis.getPageStaticInfo(params)` twice

```suggestion
      pageStaticInfo
```

**Review by timneutkens — APPROVED** [review]

**Pull request merged**